exporters/trace/stream/stream.h

```diff
@@ -0,0 +1,180 @@
+#pragma once
+
+#include "opentelemetry/sdk/trace/exporter.h"
+#include "opentelemetry/sdk/trace/span_data.h"
+#include "opentelemetry/nostd/type_traits.h"
+#include "opentelemetry/common/attribute_value.h"
+#include <opentelemetry/version.h>
+
+#include <sstream>
+#include <iostream>
+#include <map>
+
+namespace nostd    = opentelemetry::nostd;
+namespace sdktrace = opentelemetry::sdk::trace;
+
+// AttributeType to help with printing attributes
+enum AttributeType
+{
+  TYPE_BOOL,
+  TYPE_INT,
+  TYPE_INT64,
+  TYPE_UINT,
+  TYPE_UINT64,
+  TYPE_DOUBLE,
+  TYPE_STRING,
+  TYPE_SPAN_BOOL,
+  TYPE_SPAN_INT,
+  TYPE_SPAN_INT64,
+  TYPE_SPAN_UINT,
+  TYPE_SPAN_UINT64,
+  TYPE_SPAN_DOUBLE,
+  TYPE_SPAN_STRING
+};
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace exporter
+{
+namespace trace
+{
+class StreamSpanExporter final : public sdktrace::SpanExporter
+{
+public:
+    explicit StreamSpanExporter(std::ostream &sout = std::cout,
+                                bool isShutdown = false) noexcept;
+
+  std::unique_ptr<sdktrace::Recordable> MakeRecordable() noexcept override;
+
+  sdktrace::ExportResult Export(
+      const nostd::span<std::unique_ptr<sdktrace::Recordable>> &spans) noexcept override;
+
+  void Shutdown(std::chrono::microseconds timeout = std::chrono::microseconds(0)) noexcept override;
+
+private:
+  std::ostream &sout_;
+  bool isShutdown_;
+
+  // Mapping status number to the string from api/include/opentelemetry/trace/canonical_code.h
+  std::map<int, std::string> statusMap {
+    {0, "OK"},
+    {1, "CANCELLED"},
+    {2, "UNKNOWN"},
+    {3, "INVALID_ARGUMENT"},
+    {4, "DEADLINE_EXCEEDED"},
+    {5, "NOT_FOUND"},
+    {6, "ALREADY_EXISTS"},
+    {7, "PERMISSION_DENIED"},
+    {8, "RESOURCE_EXHAUSTED"},
+    {9, "FAILED_PRECONDITION"},
+    {10,"ABORTED"},
+    {11,"OUT_OF_RANGE"},
+    {12,"UNIMPLEMENTED"},
+    {13,"INTERNAL"},
+    {14,"UNAVAILABLE"},
+    {15,"DATA_LOSS"},
+    {16,"UNAUTHENTICATED"}
+  };
+
+  /*
+    print_array and print_value are used to print out the value of an attribute within a span. These values
+    are held in a variant which makes the process of printing them much more complicated.
+  */
+
+  template <typename T>
+  void print_array(common::AttributeValue &value, bool jsonTypes = false)
+  {
+    sout_ << '[';
+    // TODO: jsonTypes for bool?
+    // TODO: do we need to escape string value for JSON?
+    auto s    = nostd::get<nostd::span<const T>>(value);
+    size_t i  = 1;
+    size_t sz = s.size();
+    for (auto v : s)
+    {
+      sout_ << v; // TODO: nostd::string_view type needs quotes!
+      if (i != sz)
+        sout_ << ',';
+      i++;
+    };
+    sout_ << ']';
+  };
+
+  void print_value(common::AttributeValue &value,
+                          bool jsonTypes = false)
+  {
+    switch (value.index())
+    {
+      case AttributeType::TYPE_BOOL:
```

Comment:
The only way I currently see to safely process `nostd::variant` values is via a visitor (like e. g. [here](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/sdk/include/opentelemetry/sdk/trace/span_data.h#L36)).

I know @maxgolov has some valid concerns regarding that, and I'd appreciate him to weigh in.

With the current approach, any changes made to `common::AttributeValue` will break this code, at runtime. When using `nostd::visit`, changes to `common::AttributeValue` would cause a compiler error for this code and thus enable you to catch any incompatibilities.

Reply:
Also please rebase. `GetAttribute` on `SpanData` now returns a [SpanDataAttributeValue](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/sdk/include/opentelemetry/sdk/trace/span_data.h#L22).

Reply:
You could also do something similar to what was done [here](https://github.com/open-telemetry/opentelemetry-cpp/pull/176/files#diff-ca38b6b0446cadbf3f86cbf673da8783R24), but that doesn't warn you at compile time when the definition of the `variant` changed.

Reply:
This way of printing Attributes is from one of @maxgolov branches. Will rebase to get updated Attributes

Reply:
I rebased and am now using SpanDataAttributeValue, do you think this implementation is fine or should I go down a different path. I agree it will break if someone changes the variant, do you think that is going to happen often?

Reply:
I'd either go with [nostd::visitor](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/sdk/include/opentelemetry/sdk/trace/span_data.h#L36) or [nostd::holds_alternative](https://github.com/open-telemetry/opentelemetry-cpp/pull/176/files#diff-ca38b6b0446cadbf3f86cbf673da8783R33).

Reply:
I went with nostd::holds_alternative, let me know if you are okay with this implementation. I kept the print_array function in order to reduce code duplication.